### PR TITLE
add ioplugin for full.pem

### PR DESCRIPTION
**Pull request by mjrider**
full.pem contains key,cert and chain in 1 file, this is also a request in letsencrypt/letsencrypt#1201
this is useful for haproxy stunnel and lighttpd and other applications which expect 1 file with all the information

### CODE REVIEW

**Comment by Lekensteyn** [comment]
Isn't this already possible with `FullChainFile`?

**Comment by mjrider** [comment]
@Lekensteyn the fullchain.pem contains all the certificates, but not the private key

File: simp_le.py [hunk 1]
```diff
@@ -391,6 +391,33 @@ def save(self, data):  # pylint: disable=missing-docstring
         return super(FullChainFile, self).save(self.Data(
             key=data.key, cert=None, chain=([data.cert] + data.chain)))
 
+@IOPlugin.register(path='full.der', typ=OpenSSL.crypto.FILETYPE_ASN1)
+@IOPlugin.register(path='full.pem', typ=OpenSSL.crypto.FILETYPE_PEM)
+class FullFile(OpenSSLIOPlugin):
+    """Certificate and privatekey and chain lugin."""
```
Comment (Lekensteyn) [inline simp_le.py]: typo: lugin -> plugin, maybe be clearer about the order of the entries in the file?

```
"""Key plus full certificate chain file plugin."""
```

File: simp_le.py [hunk 2]
```diff
@@ -391,6 +391,33 @@ def save(self, data):  # pylint: disable=missing-docstring
...
+
+    def persisted(self):  # pylint: disable=missing-docstring
+        return self.Data(key=True, cert=True, chain=True)
+
+    def load(self):  # pylint: disable=missing-docstring
+        with open(self.path, 'rb') as full_file:
+            output = full_file.read().split(self._SEP)
+        key = self.load_key(output.pop(0))
+        cert = self.load_cert(output.pop(0))
+        chain = ([self.load_cert(cert_data) for cert_data in output])
```
Comment (Lekensteyn) [inline simp_le.py]: Superfluous parentheses

File: simp_le.py [hunk 3]
```diff
@@ -391,6 +391,33 @@ def save(self, data):  # pylint: disable=missing-docstring
...
+
+        return self.Data(key=key, cert=cert, chain=chain)
+
+    def save(self, data):  # pylint: disable=missing-docstring
+        logger.info('Saving %s', self.path)
+        output = [(self.dump_key(data.key))]
```
Comment (Lekensteyn) [inline simp_le.py]: again superfluous parentheses

**Comment by kuba** [comment]
I've integrated the PR with latest master in https://github.com/kuba/simp_le/tree/full-plugin.

**Comment by kuba** [comment]
This is now part of `master`

**Pull request merged**